[WPE] Add an Android platform display

aperezdc opened the pull request:
#### 42598486e6ea5601655b984c536e8f9197a369e7
<pre>
[WPE] Add an Android platform display
<a href="https://bugs.webkit.org/show_bug.cgi?id=301939">https://bugs.webkit.org/show_bug.cgi?id=301939</a>

Reviewed by Carlos Garcia Campos.

Add a new PlatformDisplayAndroid that tries using EGL_PLATFORM_ANDROID_KHR,
and also enable PlatformDisplayDefault as fallback. The fallback is needed
because it is common for the corresponding EGL_KHR_platform_android extension
to not be available, and also that some devices may not provide EGL 1.5,
which is required by the extension.

Loosely based on an initial patch by Felipe Erias &lt;felipeerias@igalia.com&gt;.

* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/android/PlatformDisplayAndroid.cpp: Added.
(WebCore::PlatformDisplayAndroid::create):
(WebCore::PlatformDisplayAndroid::PlatformDisplayAndroid):
* Source/WebCore/platform/graphics/android/PlatformDisplayAndroid.h: Added.
* Source/WebCore/platform/graphics/egl/PlatformDisplayDefault.cpp: Build
for Android.
* Source/WebCore/platform/graphics/egl/PlatformDisplayDefault.h: Ditto.
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp: Add the
correct header inclusion for drm_fourcc.h on Android.
* Source/WebKit/WebProcess/WebPage/glib/WebPageGLib.cpp:
(WebKit::WebPage::getRenderProcessInfo): Handle the Android display
type, and extend an USE(GBM) guard to account for display buffer formats
not yet being exposed on Android.
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::initializePlatformDisplayIfNeeded const): Try using
the newly added PlatformDisplayAndroid, enable trying PlatformDisplayDefault
as last resort.

Canonical link: <a href="https://commits.webkit.org/302638@main">https://commits.webkit.org/302638@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/077d05589fd4ba6a294d8be821ccc4c4b6e5164a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129643 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1900 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137031 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81100 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131514 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1833 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1776 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98757 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66605 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132590 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1423 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116116 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79429 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1340 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34247 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80303 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109815 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34746 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139513 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1704 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107274 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1762 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112457 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107127 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27296 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1400 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30968 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54447 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1761 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65124 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1585 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1620 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1695 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/android/PlatformDisplayAndroid.cpp:
```diff
@@ -0,0 +1,70 @@
+    : PlatformDisplay(WTFMove(glDisplay))
+{
+#if ENABLE(WEBGL)
+    m_anglePlatform = 0;
```
This should be EGL_PLATFORM_ANDROID_KHR when the display was created with that one. We need to make sure angle uses the same call to create the display to ensure the same eglDisplay is returned.

carlosgcampos commented on Source/WebCore/platform/graphics/android/PlatformDisplayAndroid.cpp:
```diff
@@ -0,0 +1,70 @@
+        glDisplay = GLDisplay::create(eglGetPlatformDisplay(EGL_PLATFORM_ANDROID_KHR, EGL_DEFAULT_DISPLAY, nullptr));
+#endif
+
+    glDisplay = GLDisplay::create(eglGetDisplay(EGL_DEFAULT_DISPLAY));
```
If you also enable the default display like you did in previous patch, we can leave this as android only, and the caller can fallback to the default like gtk port does.

aperezdc replied:
Okay, will do.

felipeerias committed:
[WPE] Add an Android platform display
https://bugs.webkit.org/show_bug.cgi?id=301939

Reviewed by Carlos Garcia Campos.

Add a new PlatformDisplayAndroid that tries using EGL_PLATFORM_ANDROID_KHR,
and also enable PlatformDisplayDefault as fallback. The fallback is needed
because it is common for the corresponding EGL_KHR_platform_android extension
to not be available, and also that some devices may not provide EGL 1.5,
which is required by the extension.

Loosely based on an initial patch by Felipe Erias <felipeerias@igalia.com>.

* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.h:
* Source/WebCore/platform/graphics/android/PlatformDisplayAndroid.cpp: Added.
(WebCore::PlatformDisplayAndroid::create):
(WebCore::PlatformDisplayAndroid::PlatformDisplayAndroid):
* Source/WebCore/platform/graphics/android/PlatformDisplayAndroid.h: Added.
* Source/WebCore/platform/graphics/egl/PlatformDisplayDefault.cpp: Build
for Android.
* Source/WebCore/platform/graphics/egl/PlatformDisplayDefault.h: Ditto.
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp: Add the
correct header inclusion for drm_fourcc.h on Android.
* Source/WebKit/WebProcess/WebPage/glib/WebPageGLib.cpp:
(WebKit::WebPage::getRenderProcessInfo): Handle the Android display
type, and extend an USE(GBM) guard to account for display buffer formats
not yet being exposed on Android.
* Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
(WebKit::WebProcess::initializePlatformDisplayIfNeeded const): Try using
the newly added PlatformDisplayAndroid, enable trying PlatformDisplayDefault
as last resort.

Canonical link: https://commits.webkit.org/302638@main
